clarity-ext-scripts/clarity_ext_scripts/covid/rtpcr_analysis_service.py

```diff
@@ -1,37 +1,107 @@
-    """
-    A service to analyze if a sample is positive for Covid19
-    Given:
-    CT value of sample
-    Discriminator for assay (lower bound, upper bound)
-    It's expected to be 2 controls on a plate, and these are to be
-    validated before ordinary samples are analyzed.
-    """
+
+    HEX_THRESHOLD = 32
+    FAM_THRESHOLD = 38
+
     def __init__(self):
-        self.has_valid_pos_control = False
-        self.has_valid_neg_control = False
+        pass
+
+# Understanding of the analysis criterias
+# FAM = COVID, HEX = Human internal control
+# Positive if CT(FAM) = <=38
+# Failed  if CT(FAM)= >38
+# Negative if CT(FAM) = 0 and CT(HEX) = <=32
+# Retest if CT(FAM) = 0 and CT(HEX) = >32
+
+# TODO Should we issue a re-test status here
+# TODO Later we might have to account for giving two trails over 38 as positive.
+
+    def _analyze_sample(self, sample):
+        fam_ct = sample["FAM-CT"]
+        hex_ct = sample["HEX-CT"]
+        if fam_ct > self.FAM_THRESHOLD:
+            return COVID_RESPONSE_FAILED
+        elif fam_ct == 0 and hex_ct <= self.HEX_THRESHOLD:
+            return COVID_RESPONSE_NEGATIVE
+        elif fam_ct == 0 and hex_ct > self.HEX_THRESHOLD:
+            # TODO Should retest and failed be the same thing or not.
+            return COVID_RESPONSE_FAILED
+        elif fam_ct <= self.FAM_THRESHOLD:
+            return COVID_RESPONSE_POSITIVE
+        else:
+            # TODO Don't know if it is actually a good idea to raise here or not...
+            raise NotImplementedError(
+                "Got CT-value for FAM: {} and CT-value for HEX: {}.".format(fam_ct, hex_ct))
 
-    def validate_control_value(self, control_type, ct_value, lower_bound, upper_bound):
-        if control_type.lower() == RT_PCR_POSITIVE_CONTROL and ct_value <= lower_bound:
-            self.has_valid_pos_control = True
-        if control_type.lower() == RT_PCR_NEGATIVE_CONTROL and ct_value >= upper_bound:
-            self.has_valid_neg_control = True
+    def _analyze_controls(self, positive_controls, negative_controls):
+        errors = []
+        for pos_control in positive_controls:
+            if self._analyze_sample(pos_control) == COVID_RESPONSE_NEGATIVE:
+                errors += PositiveControlWasNegative(
+                    "Positive control sample: {} was negative for covid-19".format(pos_control["id"]))
 
-    def is_valid(self):
-        return self.has_valid_neg_control and self.has_valid_pos_control
+        for neg_control in negative_controls:
+            if self._analyze_sample(neg_control) == COVID_RESPONSE_POSITIVE:
+                errors += NegativeControlWasPositive(
+                    "Negative control sample: {} was positive for covid-19".format(
+                        neg_control["id"]))
 
-    def analyze(self, ct_value, lower_bound, upper_bound):
+        if errors:
+            for e in errors:
+                log.error(e.message)
+            raise MultipleAnalysisErrors(errors)
```

Comment:
I believe we would definitely want to add this to the step error log in the extension. But it doesn't matter if it's an exception or a return status from that point of view.

Reply:
We add this to the script that uses the service.